ipfshttpclient/client/__init__.py

```diff
@@ -254,19 +260,55 @@ def add_bytes(self, data: bytes, **kwargs):
 		----------
 		data
 			Content to be added as a file
+		trickle
+			Use trickle-dag format (optimized for streaming) when generating
+			the dag; see `the old FAQ <https://github.com/ipfs/faq/issues/218>`_
+			for more information
+		only_hash
+			Only chunk and hash, but do not write to disk
+		chunker
+			The chunking algorithm to use
+		pin
+			Pin this object when adding
+		raw_leaves
+			Use raw blocks for leaf nodes. (experimental). (Default: ``True``
+			when *nocopy* is True, or ``False`` otherwise)
+		cid_version
+			CID version. Default value is provided by IPFS daemon. (experimental)
 
 		Returns
 		-------
 			str
 				Hash of the added IPFS object
 		"""
+
+		opts = {
+			"trickle": trickle,
+			"only-hash": only_hash,
+			"pin": pin,
+			"raw-leaves": raw_leaves if raw_leaves is not None else nocopy,
+		}  # type: ty.Dict[str, ty.Union[str, bool]]
+		for option_name, option_value in [
+			("chunker", chunker),
+			("cid-version", cid_version),
+		]:
+			if option_value is not None:
+				opts[option_name] = option_value
+		kwargs.setdefault("opts", {}).update(opts)
```

Comment:
I copied the implementation from `files.py`, so presumably it works. presumably what you ask can be done, I don't know how to do it. But if you provide the code I am happy to copy/paste.

Reply:
It "works" because changes to the `kwargs` `dict` - when declared as `**kwargs` - don't mutate the values of the variables at the call site in the function that called the function declaring `**kwargs` - however if it was declared as `kwargs` (no asterisks) you would mutate the call site's dictionary.

It's bad style - mutating mutable arguments can cause sneaky defects in functions that call the function doing the mutating (unless the documentation or name of the function states or strongly suggests that's the expected behavior).

Reply:
Note that if you refactor this into a new function that returns `dict` (as mentioned in another comment) the issue of `kwargs` being mutated disappears.

Reply:
@c0llab0rat0r: I don't believe that mutating the `kwargs` is a problem, since it is always a shallow copy of the caller's provided arguments.

I've seen plenty of code doing things like:

```py
def func(a, *a, **kw):
    xtra_arg = kw.pop("xtra", "default")
    ...
```

This was/is a very common trick to emulate kw-only args with writing Python 2 compatible code. Until we dropped Python 2 support py-ipfs-api-client had similar code itself…

*Note*: `kwargs` is the *exception* here, not the rule. You are right that *in general* mutating arguments like this is indeed very bad unless the function/method is clearly meant to do so.